Engine/Helper.cs

```diff
@@ -3555,4 +3555,239 @@ public object VisitUsingExpression(UsingExpressionAst usingExpressionAst)
+        }
+
+        /// <summary>
+        /// Add a vertex to the graph
+        ///
+        /// If the vertex is null, it will throw an ArgumentNullException.
+        /// If the vertex is non-null but already present in the graph, it will throw an ArgumentException
+        /// </summary>
+        /// <param name="vertex"></param>
+        public void AddVertex(T vertex)
+        {
+            ValidateNotNull(vertex);
+            if (GetIndex(vertex) != -1)
+            {
+                throw new ArgumentException(
+                    String.Format(
+                        Strings.DigraphVertexAlreadyExists,
+                        vertex),
+                    "vertex");
+            }
+
+            vertexIndexMap.Add(vertex, graph.Count);
+            graph.Add(new List<int>());
+        }
+
+        /// <summary>
+        /// Add an edge from one vertex to another
+        ///
+        /// If any input vertex is null, it will throw an ArgumentNullException
+        /// If an edge is already present between the given vertices, it will throw an ArgumentException
+        /// </summary>
+        /// <param name="fromVertex"></param>
+        /// <param name="toVertex"></param>
+        public void AddEdge(T fromVertex, T toVertex)
+        {
+            ValidateVertexArgument(fromVertex);
+            ValidateVertexArgument(toVertex);
+
+            var toIdx = GetIndex(toVertex);
+            var fromVertexList = graph[GetIndex(fromVertex)];
+            if (fromVertexList.Contains(toIdx))
+            {
+                throw new ArgumentException(String.Format(
+                    Strings.DigraphEdgeAlreadyExists,
+                    fromVertex.ToString(),
+                    toVertex.ToString()));
+            }
+            else
+            {
+                fromVertexList.Add(toIdx);
+            }
+        }
+
+        /// <summary>
+        /// Checks if a vertex is connected to another vertex within the graph
+        /// </summary>
+        /// <param name="vertex1"></param>
+        /// <param name="vertex2"></param>
+        /// <returns></returns>
+        public bool IsConnected(T vertex1, T vertex2)
+        {
+            ValidateVertexArgument(vertex1);
+            ValidateVertexArgument(vertex2);
+
+            var visited = new bool[graph.Count];
+            return IsConnected(GetIndex(vertex1), GetIndex(vertex2), ref visited);
+        }
+
+        /// <summary>
+        /// Check if two vertices are connected
+        /// </summary>
+        /// <param name="fromIdx">Origin vertex</param>
+        /// <param name="toIdx">Destination vertex</param>
+        /// <param name="visited">A boolean array indicating whether a vertex has been visited or not</param>
+        /// <returns>True if the vertices are conneted, otherwise false</returns>
+        private bool IsConnected(int fromIdx, int toIdx, ref bool[] visited)
+        {
+            visited[fromIdx] = true;
```

Comment:
I understand you are doing the marking here..but i dont see where the cleanup happens.